### PR TITLE
New module: GEM3/gem3mapper

### PR DESCRIPTION
**Pull request by miqalvarez**
As added in #3837 there's the second part of the module! The mapper part. This will be added in Sarek Pipeline soon.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by maxulysse** [comment]
Looking good.
Any way to add info about read groups?

**Comment by SPPearce** [comment]
> Looking good. Any way to add info about read groups?

The README says:
```
    -r, --sam-read-group-header=STRING
      Set read group header line (e.g. '@RG\tID:xx\tSM:yy')
      [default=none]
```

**Comment by miqalvarez** [comment]
I'll check with my supervisor and give an answer asap

File: modules/nf-core/gem3/gem3mapper/main.nf [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    path "versions.yml"           , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    gem-mapper -F 'SAM' -I $index -i $fastq -t $task.cpus | samtools view -bS -o ${prefix}.bam -
```
Comment (FriederikeHanssen) [inline modules/nf-core/gem3/gem3mapper/main.nf:25]: Sometimes people may not want to sort, can you model this in a similar way to how we do it in bwa: https://github.com/nf-core/modules/blob/5709153c7abb192d48765693f8c5a642a35e100e/modules/nf-core/bwa/mem/main.nf#L35 ?

Comment (miqalvarez) [reply modules/nf-core/gem3/gem3mapper/main.nf:25]: Sure! I'll change it, you're right

**Review by FriederikeHanssen — DISMISSED** [review]
Hey! 

I just checked the version in the mulled container and it looks like gem-mapper is on 3.6.0 and samtools on 1.6. Can you update the mulled container?

Also we recently changed the structure slightly to account for nf-test additions in the future. The conda packages are now specified in their own environemnt.yml. Maybe check some other module for an example :)

**Comment by FriederikeHanssen** [comment]
> Looking good. Any way to add info about read groups?

I think we should do this via ext.args as with the other mappers

**Comment by miqalvarez** [comment]
Hey! I'm very sry for the inactivity... I'm back again and this time no more long waits. I'll try to update the mulled container as I saw that some things changed.

About the read groups, I added the $args in the execution of gem3-mapper and (correct me if wrong as I don't fully understand) as an example, in the nextflow.config I need to put something like this?
```
...
    ext.args = { params.reads ? "-r ${params.reads}" : "" }
...
```

I've also seen that now some tests are in the modules folder. Should I move them to there or not yet?

**Comment by SPPearce** [comment]
How are we doing with this module?

**Comment by miqalvarez** [comment]
Well this module it's already done, and next on our agenda is integrating the VC '[SiNple](https://github.com/PaoloRibeca/SiNPle)' module, followed by implementing the workflow into the Sarek pipeline. But for this GEM3 module, it's all set (though feel free to confirm, as I'm still getting the hang of nf-core and maybe I missed something🫠)."

**Comment by ngspp** [comment]
Ok, let's get this module finished off and merged in if we can.
The newer testing framework is nf-test, those are the files inside the modules/nf-core/ folder. The files inside the tests/modules/nf-core are from the older pytest, these could be removed (and the entry in the test config removed too).

@FriederikeHanssen has a question with the container that has older versions of the tools, can they be updated?

**Comment by SPPearce** [comment]
Ok, let's get this module finished off and merged in if we can.
The newer testing framework is nf-test, those are the files inside the modules/nf-core/ folder. The files inside the tests/modules/nf-core are from the older pytest, these could be removed (and the entry in the test config removed too).

@FriederikeHanssen has a question with the container that has older versions of the tools, can they be updated?

File: tests/modules/nf-core/gem3/gem3mapper/main.nf
Comment (maxulysse) [inline tests/modules/nf-core/gem3/gem3mapper/main.nf:1]: We can delete this file

File: tests/modules/nf-core/gem3/gem3mapper/nextflow.config
Comment (maxulysse) [inline tests/modules/nf-core/gem3/gem3mapper/nextflow.config:1]: We can delete this file too

File: tests/modules/nf-core/gem3/gem3mapper/test.yml
Comment (maxulysse) [inline tests/modules/nf-core/gem3/gem3mapper/test.yml:1]: And that one as well

File: modules/nf-core/gem3/gem3mapper/tests/tags.yml
Comment (maxulysse) [inline modules/nf-core/gem3/gem3mapper/tests/tags.yml:1]: @adamrtalbot we can delete this file now too, right?

**Review by miqalvarez — COMMENTED** [review]
As I also made the gem3indexer module. Should I delete these tests files too in this PR?

**Comment by miqalvarez** [comment]
Regarding @FriederikeHanssen 's request, I'd like to add that I've already updated the biocontainers to the latest versions in previous commits.

File: modules/nf-core/gem3/gem3mapper/tests/tags.yml
Comment (SPPearce) [reply modules/nf-core/gem3/gem3mapper/tests/tags.yml:1]: No, don't delete this until tools is updated!

**Comment by SPPearce** [comment]
> As I also made the gem3indexer module. Should I delete these tests files too in this PR?

Only if you have swapped them to nf-test rather than pytest. Which you could do (and `nf-core modules create gem3/gem3indexer --migrate-pytest` will do a lot of the work) in this PR if you have the time.

**Comment by miqalvarez** [comment]
> > As I also made the gem3indexer module. Should I delete these tests files too in this PR?
> 
> Only if you have swapped them to nf-test rather than pytest. Which you could do (and `nf-core modules create gem3/gem3indexer --migrate-pytest` will do a lot of the work) in this PR if you have the time.

Sure! I'll do it today.

**Comment by SPPearce** [comment]
@miqalvarez , just bumping this. Shout if you need help.

**Comment by miqalvarez** [comment]
Done! Sry por being late, recently changed my workstation and had to set up the environment again

**Comment by miqalvarez** [comment]
Oops sry, I'll fix the errors after lunchtime 🫠

**Review by SPPearce — APPROVED** [review]

**Pull request merged**